src/Inventory.zig

```diff
@@ -2024,3 +2028,49 @@ pub fn loadFromZon(self: Inventory, zon: ZonElement) void {
 		}
 	}
 }
+
+fn toBase64(self: Inventory, allocator: NeverFailingAllocator) []const u8 {
+	var writer = BinaryWriter.init(main.stackAllocator);
+	defer writer.deinit();
+
+	self.toBytes(&writer);
+
+	const destination: []u8 = allocator.alloc(u8, std.base64.url_safe.Encoder.calcSize(writer.data.items.len));
+	return std.base64.url_safe.Encoder.encode(destination, writer.data.items);
+}
+
+fn toBytes(self: Inventory, writer: *BinaryWriter) void {
+	writer.writeVarInt(u32, @intCast(self._items.len));
+	for(self._items) |stack| {
+		stack.toBytes(writer);
+	}
+}
+
+fn fromBase64(self: Inventory, base64: []const u8) void {
+	const destination: []u8 = main.stackAllocator.alloc(u8, std.base64.url_safe.Decoder.calcSizeForSlice(base64) catch unreachable);
+	defer main.stackAllocator.free(destination);
+
+	std.base64.url_safe.Decoder.decode(destination, base64) catch unreachable;
+	var reader = BinaryReader.init(destination);
+	fromBytes(self, &reader);
+}
+
+fn fromBytes(self: Inventory, reader: *BinaryReader) void {
+	var count = reader.readVarInt(u32) catch 0;
+	for(self._items) |*stack| {
+		if(count == 0) {
+			stack.clear();
+			continue;
+		}
+		defer count -= 1;
+		stack.* = ItemStack.fromBytes(reader) catch |err| {
+			std.log.err("Failed to read item stack from bytes: {s}", .{@errorName(err)});
+			stack.clear();
+			continue;
+		};
+	}
+	for(0..count) |_| {
+		var stack = ItemStack.fromBytes(reader) catch continue;
```

Comment:
Should print an error. Just because the item is outside the range of the inventory, doesn't mean it can't be an indicator of a bug.

Reply:
This will cause hundreds of errors to appear if we shrink any Inventory, maybe a debug log or sth?

Reply:
That is exactly the point. We shouldn't mindlessly shrink inventories that the player put stuff in.
There should be no silent data loss.

Reply:
Yes, but we shouldn't mindlessly break it. It will just frustrate players if their hard-earned stack of amber is gone.
We should shuffle things if possible, or drop it if not.